Libraries/LibCompress/Deflate.h

```diff
@@ -26,160 +26,90 @@
-    }
-
-    bool eof() const override
-    {
-        if (!m_intermediate_stream.eof())
-            return false;
-
-        while (read_next_block()) {
-            if (!m_intermediate_stream.eof())
-                return false;
-        }
-
-        return true;
-    }
-
-    bool discard_or_error(size_t count) override
-    {
-        if (m_intermediate_stream.remaining() >= count) {
-            m_intermediate_stream.discard_or_error(count);
-            return true;
-        }
-
-        while (read_next_block()) {
-            if (m_intermediate_stream.remaining() >= count) {
-                m_intermediate_stream.discard_or_error(count);
-                return true;
-            }
-        }
-
-        m_error = true;
-        return false;
-    }
+        DeflateDecompressor& m_decompressor;
+        CanonicalCode m_literal_codes;
+        Optional<CanonicalCode> m_distance_codes;
+    };
 
-private:
-    void decompress_uncompressed_block() const;
-    void decompress_static_block() const;
-    void decompress_dynamic_block() const;
-    void decompress_huffman_block(CanonicalCode&, CanonicalCode*) const;
+    class UncompressedBlock {
+    public:
+        UncompressedBlock(DeflateDecompressor&, size_t);
+
+        bool try_read_more();
 
-    Vector<CanonicalCode> decode_huffman_codes() const;
-    u32 decode_run_length(u32) const;
-    u32 decode_distance(u32) const;
+    private:
+        DeflateDecompressor& m_decompressor;
+        size_t m_bytes_remaining;
+    };
 
-    void copy_from_history(u32, u32) const;
+    enum class State {
+        Idle,
+        ReadingCompressedBlock,
+        ReadingUncompressedBlock
+    };
+
+public:
+    friend CompressedBlock;
+    friend UncompressedBlock;
 
-    Vector<u8> generate_literal_length_codes() const;
-    Vector<u8> generate_fixed_distance_codes() const;
+    DeflateDecompressor(InputStream&);
+    ~DeflateDecompressor();
 
-    mutable BitStreamReader m_reader;
+    size_t read(Bytes) override;
+    bool read_or_error(Bytes) override;
+    bool discard_or_error(size_t) override;
+    bool eof() const override;
+
+    static ByteBuffer decompress_all(ReadonlyBytes);
+
+private:
+    u32 decode_run_length(u32);
+    u32 decode_distance(u32);
+    void decode_codes(CanonicalCode&, CanonicalCode&);
 
-    mutable CanonicalCode m_literal_length_codes;
-    mutable CanonicalCode m_fixed_distance_codes;
+    bool m_read_final_bock { false };
 
-    // FIXME: Theoretically, blocks can be extremly large, reading a single block could
-    //        exhaust memory. Maybe wait for C++20 coroutines?
-    bool read_next_block() const;
+    State m_state { State::Idle };
+    union {
+        CompressedBlock m_compressed_block;
+        UncompressedBlock m_uncompressed_block;
+    };
 
-    mutable bool m_read_last_block { false };
-    mutable DuplexMemoryStream m_intermediate_stream;
+    InputBitStream m_input_stream;
+    CircularDuplexStream<64 * 1024> m_output_stream;
```

Comment:
I might be misunderstanding this piece of code, but the circular buffer used in DEFLATE should only be 32K

Reply:
You are correct this should be 32KiB, no idea why I got 64KiB there.